### PR TITLE
[NNX] NNX migration prep (9/N): NNX-aware QK-Clip + checkpoint utilities

### PR DESCRIPTION
**Pull request by ecnal-cienet**
## NNX Migration Route Map

1. ✅ pure_nnx flag + init_state_fn scaffolding (PR #3427)
2. ✅ NNX sharding utilities (PR #3470)
3. ✅ TrainStateNNX + end-to-end training loop (PR #3500)
4. ✅ Sharding diagnostics + post-training bugfixes (PR #3652)
4.5. ✅ Linen↔NNX checkpoint conversion utilities (PR #3843)
5. ✅ NNX correctness fixes + vocab tiling on NNX (PR #3766)
6. ✅ NNX-native DPO
7. ✅ NNX-native MaxEngine inference (PR #3821)
8. ✅ NNX-native LoRA + GRPO (PR #3824)
9. 🔄 **[This PR]** NNX-aware QK-Clip + remaining NNX-format checkpoint utilities
9.5. ❌ NNX + AQT in MaxEngine + serve-mode reload + gpt3 prefill fix (stacked follow-up)
10. ❌ Vocab tiling `custom_vjp` for NNX
11. ❌ Set NNX defaults to `True`
12. ❌ Delete Linen-specific code paths

## Description

Closes the QK-Clip TODO and migrates three Linen-only checkpoint helpers (`standalone_checkpointer`, `generate_param_only_checkpoint`, `convert_gpt3_ckpt_from_paxml`) to NNX. Every NNX edit is gated on `config.pure_nnx`; Linen paths preserved byte-for-byte.

- `qk_clip_utils.apply_qk_clip_nnx` mutates `state.model` in place via `nnx.split` → pure-dict tree_map → `nnx.replace_by_pure_dict` → `nnx.update`.
- `train.py::train_step` dispatches to `apply_qk_clip_nnx` on NNX. TODO at the call site is gone.
- `standalone_checkpointer.add_entropy_to_checkpoint` dispatches across Linen `TrainState`, NNX `TrainStateNNX`, and post-split `nnx.State`.
- `generate_param_only_checkpoint` gets NNX `_possibly_unroll_params_nnx` + `_save_decode_checkpoint_nnx`, plus parallel LoRA decode flow.
- `convert_gpt3_ckpt_from_paxml` keystr_map uses the dict-style format `jax.tree_util.keystr` actually produces on `nnx.State`: `['optimizer']['step'].value`, `['optimizer']['opt_state'][0]['mu']<rest>.value` (dict for State, `[0]` for the optax tuple, `.value` for the Variable leaf).
- PR #3824 was merged before its bot review could be addressed. Both bugs are real and only hit the NNX GRPO path:
  1. `setup_train_loop` called `TrainStateNNX(model, optimizer, reference_model=...)` — but `__init__` only accepts `(model, optimizer)`. Fixed by setting `state.reference_model` as a sibling attribute after construction.
  2. `_train_step_nnx::diff_wrapper` closed over `state.reference_model` (an `nnx.Module`) inside `jax.value_and_grad`. Fixed by `nnx.split`-ing the reference outside and passing `ref_state` as an explicit arg, matching the existing policy-model pattern in the same function.

## Tests

- `qk_clip_test`: 7 new NNX cases (attention-type guard, MLA wq_b/wkv_b math, both intermediate shapes, no-clip-below-threshold, missing-stats resilience, Linen↔NNX numeric parity).
- `standalone_checkpointer_nnx_test` (new): 3 cases.
- `generate_param_only_checkpoint_nnx_test` (new): 3 cases (Llama-style, DeepSeek-style, LoRA delta unroll).
- Existing Linen tests untouched; `pure_nnx=False` stays default.

**Results:** 31 passed, 2 skipped (`qk_clip_test` + 2 new files + `grpo_nnx_test`).

## Stats

+999 / −232 across 9 files (2 new, 7 modified).

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/utils/standalone_checkpointer.py [hunk 1]
```diff
@@ -108,22 +118,38 @@ def checkpoint_loop(config, state=None):
...
+    else:
+      params = state.model.filter(nnx.Param) if hasattr(state.model, "filter") else state.model
+    new_mu = jax.tree_util.tree_map(lambda k: jnp.cos(1000 * k), params)
+    new_nu = jax.tree_util.tree_map(lambda k: jnp.sin(1000 * k), params)
+
+    if isinstance(state, nnx.Module):
+      opt = state.optimizer
+      opt.opt_state = (opt.opt_state[0]._replace(mu=new_mu, nu=new_nu),) + tuple(opt.opt_state[1:])
+    else:
+      state.optimizer.opt_state[0].mu = new_mu
```
[the remark this answers is not in the document]
Comment (ecnal-cienet) [reply src/maxtext/utils/standalone_checkpointer.py:144]: refined

File: src/maxtext/checkpoint_conversion/standalone_scripts/convert_gpt3_ckpt_from_paxml.py [hunk 2]
```diff
@@ -201,19 +227,27 @@ def get_layer_prefix(keystr_pax):
...
-        f"opt_states_0.{prefix_pax_opt_state}.m{keystr_pax}",
-        transform_fn,
-    )
-    # second momentum in optimizer state
-    state_map[f".opt_state.nu['params']{keystr_maxtext}"] = (
-        f"opt_states_0.{prefix_pax_opt_state}.v{keystr_pax}",
-        transform_fn,
-    )
+    if cfg.pure_nnx:
+      state_map[f".model{keystr_maxtext}.value"] = (f"mdl_vars{keystr_pax}", transform_fn)
```
[the remark this answers is not in the document]
Comment (ecnal-cienet) [reply src/maxtext/checkpoint_conversion/standalone_scripts/convert_gpt3_ckpt_from_paxml.py]: refined

File: src/maxtext/checkpoint_conversion/standalone_scripts/convert_gpt3_ckpt_from_paxml.py [hunk 3]
```diff
@@ -186,10 +201,21 @@ def convert(paxml_ckpt_path, maxtext_model_name, base_output_directory, run_name
...
-      ".opt_state.count": ("opt_states_0.no_prefix_0.count", None),
-  }
+  if cfg.pure_nnx:
+    # NNX state-tree paths after `nnx.split(TrainStateNNX)`:
+    #   model params     -> ['model']<rest>.value
+    #   adam mu / nu     -> ['optimizer']['opt_state']['mu' | 'nu']<rest>.value
+    #   step             -> ['optimizer']['step'].value
+    #   opt count        -> ['optimizer']['opt_state']['count'].value
+    state_map = {
+        ".optimizer.step.value": ("step", None),
```
[the remark this answers is not in the document]
Comment (ecnal-cienet) [reply src/maxtext/checkpoint_conversion/standalone_scripts/convert_gpt3_ckpt_from_paxml.py]: refined

File: src/maxtext/experimental/rl/grpo_trainer.py [hunk 4]
```diff
@@ -335,34 +339,244 @@ def grpo_loss_fn(model, config, data, dropout_rng, params, reference_params, is_
...
+        "GRPO + pure_nnx + gradient_accumulation_steps>1 not supported yet. "
+        "Set gradient_accumulation_steps=1 or pure_nnx=False."
+    )
+
+  state = nnx.merge(model_graphdef, state)  # Reconstruct the TrainStateNNX.
+  policy_graphdef, curr_params, rest = nnx.split(state.model, nnx.Param, ...)
+
+  def diff_wrapper(param, rest, config, data):
+    local_model = nnx.merge(policy_graphdef, param, rest, copy=True)
+    loss, aux = grpo_loss_fn_nnx(local_model, config, data, None, None, state.reference_model, is_train=True)
```
[the remark this answers is not in the document]
Comment (ecnal-cienet) [reply src/maxtext/experimental/rl/grpo_trainer.py]: refined

File: src/maxtext/experimental/rl/grpo_trainer.py [hunk 5]
```diff
@@ -542,26 +758,51 @@ def setup_train_loop(
...
+    if config.pure_nnx:
+      _create_model_partial, _ = model_creation_utils.create_nnx_abstract_model(config, mesh, devices=training_devices)
+
+      def init_state_fn():
+        nnx_model = _create_model_partial()
+        optimizer = nnx.Optimizer(nnx_model, tx, wrt=nnx.Param)
+        # Reference uses the same init seed so it starts identical to the policy.
+        reference_model = _create_model_partial()
+        # pylint: disable-next=unexpected-keyword-arg
+        return train_state_nnx.TrainStateNNX(nnx_model, optimizer, reference_model=reference_model)
```
[the remark this answers is not in the document]
Comment (ecnal-cienet) [reply src/maxtext/experimental/rl/grpo_trainer.py]: refined

**Review by bvandermoon — APPROVED** [review]

**Review by NuojCheng — APPROVED** [review]
